Improve test coverage for CDXML cmdlet infrastructure

JamesWTruher opened the pull request:
address issue #4159  

Create custom cim classes (via MOF) and cdxml cmdlets against the new cim class
Create tests for CRUD operations for CDXML cmdlets
Coverage for Microsoft.PowerShell.Cmdletization namespace is nearly 50%

JamesWTruher committed:
Simple CDXML tests for Get and Remove

Includes MOF files to create and delete class and instances which are used by the tests

JamesWTruher committed:
Additional tests - All CRUD operations are now covered

updated create mof file to support associations as well as complex objects
updated delete mof file to remove all classes
added an enum to cdxml (not yet used)
updated psd1 to export new cmdlets for set and new

JamesWTruher committed:
Add tests for exceptions and support whatif

adityapatwardhan commented on test/powershell/engine/Cdxml/Cdxml.Tests.ps1:
```diff
@@ -0,0 +1,268 @@
+    Context "Remove-CimTest cmdlet" {
+        BeforeEach {
+            Get-CimTest | Remove-CimTest
+            1..4 | %{ New-CimInstance -namespace root/default -class PSCore_Test1 -property @{  
```
Please use ForEach-Object instead of %

daxian-dbw replied:
This needs to be addressed as we need to the PSScriptAnalyzer warnings and errors (#3791).

JamesWTruher replied:
fixed

adityapatwardhan commented on test/powershell/engine/Cdxml/Cdxml.Tests.ps1:
```diff
@@ -0,0 +1,268 @@
+                field2 = 0
+            }
+            New-CimTest @instanceArgs
+            $result = Get-CimInstance -namespace root/default -class PSCore_Test1 | ?{$_.id -eq "telephone"}
```
Use Where-Object instead of ?

daxian-dbw replied:
Same here.

JamesWTruher replied:
fixed

adityapatwardhan commented on test/powershell/engine/Cdxml/Cdxml.Tests.ps1:
```diff
@@ -0,0 +1,268 @@
+            $result.field2 | should be 33
+        }
+    }
+
```
Extra line.

adityapatwardhan commented on test/powershell/engine/Cdxml/assets/CimTest/CdxmlTest.psd1:
```diff
@@ -0,0 +1,17 @@
+    HelpInfoUri = "https://go.microsoft.com/fwlink/?linkid=390832"
+}
+
+
```
Remove extra lines.

JamesWTruher replied:
fixed

adityapatwardhan commented on test/powershell/engine/Cdxml/assets/CimTest/CdxmlTest.psd1:
```diff
@@ -0,0 +1,17 @@
+    AliasesToExport = @()
+    CmdletsToExport = @()
+    FunctionsToExport = @( 'Get-CimTest', 'Remove-CimTest', 'New-CimTest', 'Set-CimTest' )
+    HelpInfoUri = "https://go.microsoft.com/fwlink/?linkid=390832"
```
Do we need this? This probably points to something else.

JamesWTruher replied:
probably not - removed

adityapatwardhan reviewed (approved):
LGTM with minor comments

dantraMSFT commented on test/powershell/engine/Cdxml/Cdxml.Tests.ps1:
```diff
@@ -0,0 +1,268 @@
+Describe "Cdxml cmdlets are supported" -Tag CI,RequireAdminOnWindows {
+    BeforeAll {
+        $skipNotWindows = ! $IsWindows
+        if ( $skipNotWindows ) {
```
Why not use a direct $IsWindows test versus a variable?

JamesWTruher replied:
I thought it was clearer here `-skip` (I've seen other tests do exactly the opposite of what was wanted)

dantraMSFT commented on test/powershell/engine/Cdxml/Cdxml.Tests.ps1:
```diff
@@ -0,0 +1,268 @@
+            }
+            $result = MofComp.exe $deleteMof
+            if ( $LASTEXITCODE -ne 0 ) {
+                $script:ItSkipOrPending = @{ Pending = $true }
```
Why isn't this an outright failure? It appears to be masking a bug.

JamesWTruher replied:
I thought about this, and while it might be a bug in the mof file or a problem with mofcomp, marking a test as pending doesn't get us off the hook - all pending tests must be resolved before we ship (and I do track our pending tests). However, I didn't want to invalidate the build because I couldn't even get to the actual test. I thought it was better to mark the test as pending rather than fail.

dantraMSFT commented on test/powershell/engine/Cdxml/Cdxml.Tests.ps1:
```diff
@@ -0,0 +1,268 @@
+            $script:ItSkipOrPending = @{ Pending = $true }
+            return
+        }
+        # now load the cdxml module
```
should be 'unload the cdxml module'

dantraMSFT commented on test/powershell/engine/Cdxml/Cdxml.Tests.ps1:
```diff
@@ -0,0 +1,268 @@
+                $job = Get-CimTest -id 3 | Remove-CimTest -asjob
+                $result = $null
+                $i = 0
+                # wait up to 10 seconds, then the test will fail
```
Why isn't Wait-Job -Timeout sufficient?

JamesWTruher replied:
ahem, yes, that's much better
-fixed-

JamesWTruher committed:
Add hardening to test code to handle missing or misbehaving MOFCOMP

If there is a problem with mofcomp execution, the tests will fail rather than
being marked as pending (feedback from DanTra)
use cmdlet names rather than aliases for foreach-object and where-object
Script analyzer report is clean

Also remove unneeded empty lines and helpinfouri from psd1 file

JamesWTruher commented:
@dantraMSFT, I believe this is ok now

daxian-dbw commented:
@dantraMSFT Can you please take another look at this PR?
